rules/group.go

```diff
@@ -167,6 +200,78 @@ func (g *Group) Limit() int { return g.limit }
 
 func (g *Group) Logger() log.Logger { return g.logger }
 
+func (g *Group) reset() {
+	n := GroupKey(g.file, g.name)
+	if m := g.metrics; m != nil {
+		m.GroupLastEvalTime.DeleteLabelValues(n)
+		m.GroupLastDuration.DeleteLabelValues(n)
+		m.GroupSamples.DeleteLabelValues(n)
+
+		m.GroupLastEvalTime.WithLabelValues(n)
+		m.GroupLastDuration.WithLabelValues(n)
+		m.GroupSamples.WithLabelValues(n)
+	}
+
+	for _, rule := range g.Rules() {
+		rule.SetHealth(HealthUnknown)
+		rule.SetEvaluationTimestamp(time.Time{})
+		rule.SetEvaluationDuration(0)
+		rule.SetLastError(nil)
+
+		alertRule, ok := rule.(*AlertingRule)
+		if !ok {
+			continue
+		}
+		alertRule.active = map[uint64]*Alert{}
+		alertRule.SetRestored(false)
+	}
+	g.setLastEvaluation(time.Time{})
+	g.setEvaluationTime(0)
+	g.setLastEvalTimestamp(time.Time{})
+	g.setShouldRestore(true)
+
+	g.seriesInPreviousEval = make([]map[string]labels.Labels, len(g.Rules()))
+	g.staleSeries = nil
+}
+
+func (g *Group) evaluateRuleGroup(ctx context.Context, evalTimestamp time.Time, missed time.Duration) {
+	if missed > 0 {
+		g.metrics.IterationsMissed.WithLabelValues(GroupKey(g.file, g.name)).Add(float64(missed))
+		g.metrics.IterationsScheduled.WithLabelValues(GroupKey(g.file, g.name)).Add(float64(missed))
+	}
+	if g.shouldEvaluateRuleGroupFunc(ctx, g, evalTimestamp, g.done) {
+		g.setState(GroupStateActive)
+		g.evalIterationFunc(ctx, g, evalTimestamp)
+		return
+	}
+	g.reset()
+	g.setState(GroupStateInactive)
+}
+
+func (g *Group) ShouldRestore() bool {
```

Comment:
I see that some of these methods are exported (capitalized), others are not. Are the exported ones used from the outside by your code? Otherwise I'd keep them all unexported to make it clear that that's not the case.

And just to check that I am understanding things correctly, are these mutex-protected accessor functions only needed for testing, not because of concurrency protection within the rule manager itself?

Reply:
I am changing this to `requiresRestoration()`. I cannot keep the field and method name the same 

> And just to check that I am understanding things correctly, are these mutex-protected accessor functions only needed for testing, not because of concurrency protection within the rule manager itself?

Correct, needed only for testing